### PR TITLE
[ML] Stratified cross validation split for classification

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
As classification now works for multiple classes, randomly
picking training/test data frame rows is not good enough.
This commit introduces a stratified cross validation splitter
that maintains the proportion of the each class in the dataset
in the sample that is used for training the model.

### CODE REVIEW

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+            }
+        }
+
+        classCardinalities = new HashMap<>();
+        classCardinalities.put("a", classA);
+        classCardinalities.put("b", classB);
+        classCardinalities.put("c", classC);
+    }
+
+    public void testConstructor_GivenMissingDependendVariable() {
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java]: s/DependendVariable/DependentVariable/

**Review by benwtrent — APPROVED** [review]
Looks great!

Two minor concerns :)

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+            String[] row = new String[fields.size()];
+            for (int fieldIndex = 0; fieldIndex < fields.size(); fieldIndex++) {
+                String value = fieldIndex == dependentVariableIndex ? classValuesPerRow[i] : randomAlphaOfLength(10);
+                row[fieldIndex] = value;
+            }
+
+            String[] processedRow = Arrays.copyOf(row, row.length);
+            splitter.process(processedRow, this::incrementTrainingDocsCount, this::incrementTestDocsCount);
+
+            // As all these rows have no dependent variable value, they're not for training and should be unaffected
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java]: Is this comment relevant to this test case?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/CrossValidationSplitterFactory.java [hunk 3]
```diff
@@ -5,32 +5,84 @@
...
+            return createStratifiedSplitter((Classification) config.getAnalysis());
         }
         return (row, incrementTrainingDocs, incrementTestDocs) -> incrementTrainingDocs.run();
     }
+
+    private CrossValidationSplitter createStratifiedSplitter(Classification classification) {
+        String aggName = "dependent_variable_terms";
+        SearchRequestBuilder searchRequestBuilder = client.prepareSearch(config.getDest().getIndex())
+            .setSize(0)
+            .addAggregation(AggregationBuilders.terms(aggName)
```
[the remark this answers is not in the document]
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/CrossValidationSplitterFactory.java]: Definitely!

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+                if (processedRow[dependentVariableIndex] != DataFrameDataExtractor.NULL_VALUE) {
+                    trainingCountPerRow[i]++;
+                }
+            }
+        }
+
+        // We expect each data row to be selected uniformly.
+        // Thus the fraction of the row count where it's selected for training against the number of runs
+        // should be close to the training percent, which is set to 0.5
```
Comment (przemekwitek) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java]: But what if we are unlucky with random generator?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/CrossValidationSplitterFactory.java [hunk 5]
```diff
@@ -5,32 +5,84 @@
...
+            throw new ElasticsearchException("[" + config.getId() + "] Dependent variable terms search encountered ["
+                + unavailableShards + "] unavailable shards");
+        }
+
+        Aggregations aggs = searchResponse.getAggregations();
+        Terms terms = aggs.get(aggName);
+        Map<String, Long> classCardinalities = new HashMap<>();
+        for (Terms.Bucket bucket : terms.getBuckets()) {
+            classCardinalities.put(String.valueOf(bucket.getKey()), bucket.getDocCount());
+        }
```
[the remark this answers is not in the document]
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/CrossValidationSplitterFactory.java]: Yes. Also, you noted that I ended up adding a completely new search for getting the class counts. I decided not to complicate all that code for performance that is probably negligible compared to running the analysis itself. We can always optimize when we have evidence it's critical.

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java (hunk 2 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java]: Oups

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java (hunk 4 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java]: I have made several runs and plotted the distribution to see how it looks like. We get plots like this:

image

So the 0.1 margin is quite roomy. If the test fails then we'll learn this doesn't perform as we thought it was.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsProcessManager.java [hunk 6]
```diff
@@ -324,7 +324,8 @@ private void refreshIndices(String jobId) {
         );
         refreshRequest.indicesOptions(IndicesOptions.lenientExpandOpen());
 
-        LOGGER.debug("[{}] Refreshing indices {}", jobId, Arrays.toString(refreshRequest.indices()));
+        LOGGER.debug(() -> new ParameterizedMessage("[{}] Refreshing indices {}",
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/process/AnalyticsProcessManager.java]: @benwtrent I forgot to address your comment before so I squeezed this one here

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java (hunk 4 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java]: I love graphs :D

Comment (przemekwitek) [reply x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/dataframe/process/crossvalidation/StratifiedCrossValidationSplitterTests.java]: Ack

**Review by przemekwitek — APPROVED** [review]
LGTM
